### ISSUE TITLE
Regression: (seizure risk?) screen flickers rapidly when using tracy

### ISSUE THREAD

**Issue by aevyrie** [bevyengine/bevy#22257]
Running the scene viewer with caldera hotel (only tested on Mac) flickers violently when using the trace-tracy feature.

## Minimal Repro

```
cargo run --example bloom_3d --features=trace_tracy
```

**Comment by mockersf** [comment]
I could not reproduce on my mac with `cargo run --example load_gltf --features trace_tracy` or `many_foxes`

**Comment by aevyrie** [comment]
it seems to happen on more challenging scenes, like caldera.

**Comment by mockersf** [comment]
I can reproduce on https://github.com/DGriffin91/bevy_caldera_scene

On that scene if I don't enable `Bloom` and `ScreenSpaceAmbientOcclusion`, it doesn't flicker even with `bevy/trace_tracy` enabled

**Comment by alice-i-cecile** [comment]
I think that this is likely due to the bug with schedule ordering that we fixed post 0.18.rc-1: #21930. This sounds like a critical system ordering is getting screwed up, but only when the scheduler is under enough load to actually trigger a flip-flop.

It would be good to try and reproduce this bug again.

**Comment by mockersf** [comment]
It doesn't need real load, just rendering passes. It can be reproduced on the `bloom_3d` example

**Comment by aevyrie** [comment]
I had griffin's bistro repo all set up to point to my local bevy files and retested on main, still seeing the awful flashing.

# WARNING: FLASHING/STROBING LIGHTS

https://github.com/user-attachments/assets/c959daca-3714-4249-98e0-3fc87143d393

**Comment by aevyrie** [comment]
Can confirm I can repro in repo with just
```
cargo run --example bloom_3d --features=trace_tracy
```

**Comment by hukasu** [comment]
ran `bloom_3d` on 730b0022604ff2b09ac80cf385c1a38050203f41 and only blinks black once on new pipeline instantiations when changing some of the parameters

Command and system info
```
/mnt/data/Code/bevy_fork> cargo run --example bloom_3d --features="trace_tracy"
   Compiling bevy v0.18.0-dev (/mnt/data/Code/bevy_fork)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 3.08s
     Running `target/debug/examples/bloom_3d`
2026-01-01T22:16:14.594915Z  WARN bevy_log: Tracing with Tracy is active, memory consumption will grow until a client is connected
2026-01-01T22:16:14.598725Z  INFO plugin build{plugin="bevy_diagnostic::DiagnosticsPlugin"}: bevy_diagnostic::system_information_diagnostics_plugin::internal: SystemInfo { os: "Linux (NixOS 25.11)", kernel: "6.12.63", cpu: "Intel(R) Core(TM) i7-14700KF", core_count: "20", memory: "31.2 GiB" }
2026-01-01T22:16:14.662756Z  INFO plugin build{plugin="bevy_render::RenderPlugin"}: bevy_render::renderer: AdapterInfo { name: "NVIDIA GeForce RTX 4060 Ti", vendor: 4318, device: 10243, device_type: DiscreteGpu, driver: "NVIDIA", driver_info: "580.119.02", backend: Vulkan }
2026-01-01T22:16:15.264180Z  INFO bevy_app:plugin finish:plugin finish{plugin="bevy_render::batching::gpu_preprocessing::BatchingPlugin"}: bevy_render::batching::gpu_preprocessing: GPU preprocessing is fully supported on this device.
```

have been opening multiple instances of `bloom3d` and still haven't seen it happen on linux

**Comment by kfc35** [comment]
ran `cargo run --example bloom_3d --features=trace_tracy` on main several times (and also 6 instances deep at times) and there was no flickering for me. Apple M4

```
2026-01-01T23:54:42.395241Z  WARN bevy_log: Tracing with Tracy is active, memory consumption will grow until a client is connected
2026-01-01T23:54:42.397209Z  INFO plugin build{plugin="bevy_diagnostic::DiagnosticsPlugin"}: bevy_diagnostic::system_information_diagnostics_plugin::internal: SystemInfo { os: "macOS 15.7.3 Sequoia", kernel: "24.6.0", cpu: "Apple M4", core_count: "10", memory: "16.0 GiB" }
2026-01-01T23:54:42.643603Z  INFO plugin build{plugin="bevy_render::RenderPlugin"}: bevy_render::renderer: AdapterInfo { name: "Apple M4", vendor: 0, device: 0, device_type: IntegratedGpu, driver: "", driver_info: "", backend: Metal }
2026-01-01T23:54:43.429739Z  INFO bevy_app:plugin finish:plugin finish{plugin="bevy_render::batching::gpu_preprocessing::BatchingPlugin"}: bevy_render::batching::gpu_preprocessing: GPU preprocessing is fully supported on this device.
```

**Comment by aevyrie** [comment]
This is looking like a MacOS Tahoe-only issue. I tried bisecting, but it was extremely challenge to reliable repro the further back I went. There was also a confounding jitter issue that might be unrelated. The issue seems to be caused by the gpu timing diagnostics, but again, very difficult to reproduce. Best bet is to feature gate those lines and see if it fixes the issue on main, where it is very easy to reproduce.

**Comment by hukasu** [comment]
i wonder what would show up on the XCode Gpu profiler thingy on one of the white frames

**Comment by alice-i-cecile** [comment]
Splash damage seems limited, and we're struggling to reproduce and isolate this. I think we should cut from the milestone?

**Comment by it-me-joda** [comment]
Unfortunately works on my machine and tested up to 10 instances. 
Linux and tested against Wayland and X11. 

```
2026-01-03T04:59:32.207548Z  WARN bevy_log: Tracing with Tracy is active, memory consumption will grow until a client is connected
2026-01-03T04:59:32.213107Z  INFO plugin build{plugin="bevy_diagnostic::DiagnosticsPlugin"}: bevy_diagnostic::system_information_diagnostics_plugin::internal: SystemInfo { os: "Linux (CachyOS Linux rolling)", kernel: "6.18.2-2-cachyos", cpu: "AMD Ryzen 7 3700X 8-Core Processor", core_count: "8", memory: "62.7 GiB" }
2026-01-03T04:59:32.263811Z  INFO plugin build{plugin="bevy_render::RenderPlugin"}: bevy_render::renderer: AdapterInfo { name: "AMD Radeon RX 6950 XT (RADV NAVI21)", vendor: 4098, device: 29605, device_type: DiscreteGpu, driver: "radv", driver_info: "Mesa 25.3.3-arch1.2", backend: Vulkan }
```

**Comment by kfc35** [comment]
^ just opened https://github.com/bevyengine/bevy/pull/22359 since #19191 was suspected to be the start of the source of the flickering… hope it helps. I went and just gated all of the diagnostics that I could search beyond what is in 19191

**Comment by esmevane** [comment]
Apple M1 / Sonoma, unable to reproduce here.

**Comment by hukasu** [comment]
after joda's comment i made a script (in nushell) to open 10 instances of `bloom_3d` 100 times and still haven't seen it on wayland or x11

```nu
for $x in 0..100 {
    if ($x + 1) mod 10 == 0 { print $x }
    let ids = 0..10 | each { |$y| job spawn { cargo run --example bloom_3d --features="trace_tracy" } }
    sleep 10sec
    for $id in $ids { job kill $id }
}
```

**Comment by janhohenheim** [comment]
> Splash damage seems limited

This can cause seizures in players without the developers knowing about it, so the devs wouldn't even be able to add a seizure warning. I know this is only here when using tracy, but I know plenty of devs just leave the tracy feature on in their releases to not have to deal with dev feature gates. Worst case, a player at seizure risk downloads a cozy game made with Bevy that should have no flashing lights and BAM, seizure. IMO this is critical.
Also, there's a real risk of unsuspecting devs trying to profile their stuff getting a seizure.

**Comment by aevyrie** [comment]
I narrowed this down to a single line

```rs
recorder.finish_frame
```

I wasn't able to surgically remove anything *inside* that function without causing other issues I don't understand, but so far this is a reliable fix.

**Comment by hukasu** [comment]
does changing both `cfg`s to ... solve without issues?
```rust
#[cfg(all(feature = "tracing-tracy", not(target_os = "macos"))]
```

https://github.com/bevyengine/bevy/blob/e457025cc447a9c36ac452f89c3da9ef836cd0d5/crates/bevy_render/src/diagnostic/internal.rs#L117-L118

https://github.com/bevyengine/bevy/blob/e457025cc447a9c36ac452f89c3da9ef836cd0d5/crates/bevy_render/src/diagnostic/internal.rs#L133-L134
